<a href="https://colab.research.google.com/github/KomisD/Drug-Classification/blob/main/Project2_NNs.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Importing the libraries

In [733]:
#pip install category_encoders

In [734]:
import numpy as np
import matplotlib.pyplot as plt
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split, KFold, cross_val_score
import pandas as pd
import category_encoders as ce

# Simple Support Vector Machine (SVM)

## Importing the dataset and encoding categorical values


In [735]:
dataset = pd.read_csv('drug200.csv')

# Applying the binary encoder for SEX
binary_encoder = ce.BinaryEncoder(cols=['Sex'])
dataset = binary_encoder.fit_transform(dataset)

# Applying the OneHot encoder for blood pressure
OneHot = ce.OneHotEncoder(cols=['BP'])
dataset = OneHot.fit_transform(dataset)

#Applaying binary encoder for Cholesterol
binary_encoder = ce.BinaryEncoder(cols=['Cholesterol'])
dataset = binary_encoder.fit_transform(dataset)

#spliting variables into depended and target value
X = dataset.iloc[:, :-1].values
y = dataset.iloc[:, -1].values

# Applying label encoding to the target variable 'Drug'
label_encoder = LabelEncoder()
y = label_encoder.fit_transform(y)

## Splitting the dataset into the Training set and Test set

In [736]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.4)

In [737]:
print(X_train)

[[51.     0.     1.    ...  0.     1.    13.597]
 [36.     0.     1.    ...  0.     1.    16.753]
 [64.     1.     0.    ...  0.     1.     7.761]
 ...
 [49.     1.     0.    ...  1.     0.    11.014]
 [48.     1.     0.    ...  1.     0.    10.446]
 [61.     0.     1.    ...  0.     1.    18.043]]


In [738]:
print(y_train)

[4 0 4 0 0 4 1 0 0 4 2 1 2 0 2 4 1 2 0 1 3 0 0 0 3 4 1 4 0 0 4 3 3 0 4 0 4
 4 0 4 4 0 4 0 0 0 0 1 0 3 0 0 2 0 0 4 0 0 0 4 4 4 3 3 4 2 4 4 1 0 2 4 0 0
 4 1 0 1 1 0 3 0 1 0 0 0 4 2 4 0 2 0 1 0 0 4 0 0 0 0 0 4 1 4 4 0 0 0 4 4 0
 1 0 2 0 4 0 4 1 0]


In [739]:
print(X_test)

[[68.     1.     0.     0.     1.     0.     0.     1.    10.291]
 [34.     1.     0.     0.     0.     1.     0.     1.    22.456]
 [66.     1.     0.     1.     0.     0.     0.     1.    16.347]
 [47.     0.     1.     0.     1.     0.     0.     1.    10.067]
 [39.     1.     0.     0.     0.     1.     0.     1.    15.969]
 [32.     0.     1.     0.     0.     1.     0.     1.     7.477]
 [65.     0.     1.     0.     1.     0.     1.     0.    13.769]
 [67.     1.     0.     0.     0.     1.     1.     0.    10.898]
 [56.     0.     1.     0.     1.     0.     0.     1.    11.567]
 [20.     0.     1.     0.     1.     0.     1.     0.    11.686]
 [50.     0.     1.     0.     0.     1.     0.     1.    12.703]
 [26.     0.     1.     1.     0.     0.     1.     0.    12.307]
 [69.     1.     0.     0.     1.     0.     1.     0.    11.455]
 [72.     0.     1.     0.     1.     0.     1.     0.    14.642]
 [28.     1.     0.     0.     0.     1.     0.     1.    27.064]
 [34.     

In [740]:
print(y_test)

[3 0 0 3 0 4 4 4 3 4 4 1 4 4 0 0 2 0 3 4 2 4 0 4 0 0 0 1 0 4 1 4 0 0 0 0 1
 2 4 0 0 3 0 0 0 0 3 4 1 3 0 4 0 0 0 0 0 2 0 4 0 4 2 0 0 0 1 3 4 0 0 0 4 0
 0 1 4 2 4 1]


## Feature Scaling

In [741]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

In [742]:
print(X_train)

[[ 0.39037891 -1.06904497  1.06904497 ... -1.05131497  1.05131497
  -0.30709991]
 [-0.49909203 -1.06904497  1.06904497 ... -1.05131497  1.05131497
   0.12482349]
 [ 1.16125373  0.93541435 -0.93541435 ... -1.05131497  1.05131497
  -1.10580235]
 ...
 [ 0.27178279  0.93541435 -0.93541435 ...  0.95118973 -0.95118973
  -0.66060375]
 [ 0.21248473  0.93541435 -0.93541435 ...  0.95118973 -0.95118973
  -0.73833901]
 [ 0.98335954 -1.06904497  1.06904497 ... -1.05131497  1.05131497
   0.30137012]]


In [743]:
print(X_test)

[[ 1.39844598  0.93541435 -0.93541435 -0.81649658  1.46897745 -0.62876771
  -1.05131497  1.05131497 -0.75955198]
 [-0.61768816  0.93541435 -0.93541435 -0.81649658 -0.68074565  1.59041245
  -1.05131497  1.05131497  0.90532384]
 [ 1.27984986  0.93541435 -0.93541435  1.22474487 -0.68074565 -0.62876771
  -1.05131497  1.05131497  0.0692592 ]
 [ 0.15318666 -1.06904497  1.06904497 -0.81649658  1.46897745 -0.62876771
  -1.05131497  1.05131497 -0.79020814]
 [-0.32119784  0.93541435 -0.93541435 -0.81649658 -0.68074565  1.59041245
  -1.05131497  1.05131497  0.01752693]
 [-0.73628428 -1.06904497  1.06904497 -0.81649658 -0.68074565  1.59041245
  -1.05131497  1.05131497 -1.14466998]
 [ 1.22055179 -1.06904497  1.06904497 -0.81649658  1.46897745 -0.62876771
   0.95118973 -0.95118973 -0.28356035]
 [ 1.33914792  0.93541435 -0.93541435 -0.81649658 -0.68074565  1.59041245
   0.95118973 -0.95118973 -0.67647926]
 [ 0.68686923 -1.06904497  1.06904497 -0.81649658  1.46897745 -0.62876771
  -1.05131497  1.05131

## Training the SVM model on the Training set

In [744]:
from sklearn.svm import SVC
classifier = SVC(kernel = 'linear', random_state=0)
classifier.fit(X_train, y_train)

SVC(kernel='linear', random_state=0)

## Predicting the Test set results

In [745]:
y_pred = classifier.predict(X_test)
print(np.concatenate((y_pred.reshape(len(y_pred),1), y_test.reshape(len(y_test),1)),1))

[[3 3]
 [0 0]
 [0 0]
 [3 3]
 [0 0]
 [4 4]
 [0 4]
 [4 4]
 [3 3]
 [4 4]
 [4 4]
 [1 1]
 [4 4]
 [0 4]
 [0 0]
 [0 0]
 [0 2]
 [0 0]
 [3 3]
 [4 4]
 [2 2]
 [4 4]
 [0 0]
 [4 4]
 [0 0]
 [0 0]
 [0 0]
 [1 1]
 [0 0]
 [4 4]
 [1 1]
 [4 4]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [1 1]
 [2 2]
 [4 4]
 [0 0]
 [0 0]
 [3 3]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [3 3]
 [0 4]
 [1 1]
 [3 3]
 [0 0]
 [4 4]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [2 2]
 [0 0]
 [4 4]
 [0 0]
 [4 4]
 [2 2]
 [0 0]
 [0 0]
 [0 0]
 [1 1]
 [3 3]
 [0 4]
 [0 0]
 [0 0]
 [0 0]
 [4 4]
 [0 0]
 [0 0]
 [1 1]
 [4 4]
 [2 2]
 [4 4]
 [1 1]]


## Making the Confusion Matrix

In [746]:
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_test, y_pred)
print(cm)
accuracy_score(y_test, y_pred)

[[37  0  0  0  0]
 [ 0  8  0  0  0]
 [ 1  0  5  0  0]
 [ 0  0  0  8  0]
 [ 4  0  0  0 17]]


0.9375

# Simple Support Vector Machine (SVM) with cross validation

## Importing the dataset and encoding categorical values


In [747]:
dataset = pd.read_csv('drug200.csv')

# Applying the binary encoder for SEX
binary_encoder = ce.BinaryEncoder(cols=['Sex'])
dataset = binary_encoder.fit_transform(dataset)

# Applying the OneHot encoder for blood pressure
OneHot = ce.OneHotEncoder(cols=['BP'])
dataset = OneHot.fit_transform(dataset)

#Applaying binary encoder for Cholesterol
binary_encoder = ce.BinaryEncoder(cols=['Cholesterol'])
dataset = binary_encoder.fit_transform(dataset)

#spliting variables into depended and target value
X = dataset.iloc[:, :-1].values
y = dataset.iloc[:, -1].values

# Applying label encoding to the target variable 'Drug'
label_encoder = LabelEncoder()
y = label_encoder.fit_transform(y)

## Splitting the dataset into the Training set and Test set

In [748]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.4)

In [749]:
print(X_train)

[[50.     1.     0.    ...  1.     0.    15.79 ]
 [35.     1.     0.    ...  1.     0.     9.17 ]
 [42.     1.     0.    ...  1.     0.    12.766]
 ...
 [23.     1.     0.    ...  0.     1.    31.686]
 [20.     0.     1.    ...  1.     0.    11.686]
 [74.     1.     0.    ...  1.     0.    11.939]]


In [750]:
print(y_train)

[0 4 1 0 0 4 0 0 0 0 4 3 4 2 4 0 4 0 4 4 0 0 1 0 0 0 4 3 2 0 4 4 0 0 0 4 3
 3 0 2 4 0 0 0 4 2 1 0 0 4 0 4 0 1 4 4 3 2 0 4 3 0 3 3 4 0 0 0 0 4 4 4 0 0
 4 0 1 4 0 0 1 0 1 4 2 4 0 0 0 1 0 1 2 0 1 0 0 0 0 0 2 0 0 4 0 0 0 3 3 4 0
 0 0 4 0 3 1 0 4 4]


In [751]:
print(X_test)

[[26.     0.     1.     1.     0.     0.     1.     0.    12.307]
 [30.     0.     1.     0.     0.     1.     0.     1.    10.443]
 [36.     0.     1.     1.     0.     0.     1.     0.    15.49 ]
 [24.     0.     1.     0.     0.     1.     0.     1.    10.605]
 [22.     1.     0.     0.     0.     1.     0.     1.    11.953]
 [60.     1.     0.     1.     0.     0.     1.     0.     8.621]
 [49.     1.     0.     1.     0.     0.     1.     0.     6.269]
 [31.     1.     0.     1.     0.     0.     1.     0.    17.069]
 [64.     0.     1.     0.     1.     0.     1.     0.    25.741]
 [37.     0.     1.     1.     0.     0.     0.     1.    13.091]
 [15.     1.     0.     0.     0.     1.     0.     1.     9.084]
 [69.     1.     0.     0.     1.     0.     1.     0.    11.455]
 [40.     0.     1.     0.     0.     1.     0.     1.    10.103]
 [60.     1.     0.     1.     0.     0.     0.     1.    13.934]
 [50.     0.     1.     0.     0.     1.     1.     0.    17.211]
 [25.     

In [752]:
print(y_test)

[1 4 0 4 4 2 1 0 0 1 4 4 4 2 0 0 4 0 0 0 2 4 4 0 1 1 4 1 1 0 1 4 1 4 2 0 2
 0 4 3 0 2 4 0 4 0 4 1 3 0 3 0 0 0 1 2 0 4 0 0 0 0 3 0 0 2 1 4 0 0 4 0 4 0
 3 4 0 0 0 4]


## Feature Scaling

In [753]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

In [754]:
print(X_train)

[[ 0.39974131  0.98346994 -0.98346994 ...  1.05131497 -1.05131497
  -0.1324063 ]
 [-0.55708629  0.98346994 -0.98346994 ...  1.05131497 -1.05131497
  -1.00040786]
 [-0.11056675  0.98346994 -0.98346994 ...  1.05131497 -1.05131497
  -0.52890731]
 ...
 [-1.32254838  0.98346994 -0.98346994 ... -0.95118973  0.95118973
   1.9518464 ]
 [-1.5139139  -1.0168079   1.0168079  ...  1.05131497 -1.05131497
  -0.67051482]
 [ 1.93066548  0.98346994 -0.98346994 ...  1.05131497 -1.05131497
  -0.63734195]]


In [755]:
print(X_test)

[[-1.13118286 -1.0168079   1.0168079   1.33816952 -0.74728947 -0.62876771
   1.05131497 -1.05131497 -0.5890905 ]
 [-0.87602883 -1.0168079   1.0168079  -0.74728947 -0.74728947  1.59041245
  -0.95118973  0.95118973 -0.83349457]
 [-0.49329779 -1.0168079   1.0168079   1.33816952 -0.74728947 -0.62876771
   1.05131497 -1.05131497 -0.17174171]
 [-1.25875987 -1.0168079   1.0168079  -0.74728947 -0.74728947  1.59041245
  -0.95118973  0.95118973 -0.81225344]
 [-1.38633689  0.98346994 -0.98346994 -0.74728947 -0.74728947  1.59041245
  -0.95118973  0.95118973 -0.6355063 ]
 [ 1.03762638  0.98346994 -0.98346994  1.33816952 -0.74728947 -0.62876771
   1.05131497 -1.05131497 -1.07239168]
 [ 0.3359528   0.98346994 -0.98346994  1.33816952 -0.74728947 -0.62876771
   1.05131497 -1.05131497 -1.38078136]
 [-0.81224032  0.98346994 -0.98346994  1.33816952 -0.74728947 -0.62876771
   1.05131497 -1.05131497  0.0352937 ]
 [ 1.29278041 -1.0168079   1.0168079  -0.74728947  1.33816952 -0.62876771
   1.05131497 -1.05131

## Creating the SVM model

In [756]:
from sklearn.svm import SVC
classifier = SVC(kernel = 'linear', random_state=0)

## Applying Cross Validation

In [757]:
# Define the k-fold cross-validator (k=5)
kf = KFold(n_splits=2, shuffle=True)

# Perform cross-validation
cv_scores = cross_val_score(classifier, X, y, cv=kf)

## Printing Scores

In [758]:
# Print the cross-validation scores
print("Cross-validation scores:", cv_scores)
print("Mean CV Score:", cv_scores.mean())

Cross-validation scores: [0.98 0.99]
Mean CV Score: 0.985


##Triying different n_splits

In [759]:
# Define the k-fold cross-validator (k=10)
kf = KFold(n_splits=10, shuffle=True, random_state=0)

# Perform cross-validation
cv_scores = cross_val_score(classifier, X, y, cv=kf)

# Print the cross-validation scores
print("Cross-validation scores:", cv_scores)
print("Mean CV Score:", cv_scores.mean())

Cross-validation scores: [1.   1.   0.95 0.95 1.   1.   1.   1.   1.   1.  ]
Mean CV Score: 0.99


In [760]:
# Define the k-fold cross-validator (k=15)
kf = KFold(n_splits=15, shuffle=True)

# Perform cross-validation
cv_scores = cross_val_score(classifier, X, y, cv=kf)

# Print the cross-validation scores
print("Cross-validation scores:", cv_scores)
print("Mean CV Score:", cv_scores.mean())

Cross-validation scores: [0.92857143 1.         1.         1.         1.         1.
 1.         1.         1.         1.         1.         0.92307692
 1.         1.         1.        ]
Mean CV Score: 0.9901098901098901


In [761]:
# Define the k-fold cross-validator (k=25)
kf = KFold(n_splits=25, shuffle=True)

# Perform cross-validation
cv_scores = cross_val_score(classifier, X, y, cv=kf)

# Print the cross-validation scores
print("Cross-validation scores:", cv_scores)
print("Mean CV Score:", cv_scores.mean())

Cross-validation scores: [1.    1.    1.    1.    1.    1.    1.    1.    1.    1.    1.    1.
 1.    1.    1.    1.    1.    1.    1.    1.    1.    1.    1.    0.875
 0.875]
Mean CV Score: 0.99


# Simple Support Vector Machine (SVM) with PCA Applayied

## Importing the dataset and encoding categorical values


In [762]:
dataset = pd.read_csv('drug200.csv')

# Applying the binary encoder for SEX
binary_encoder = ce.BinaryEncoder(cols=['Sex'])
dataset = binary_encoder.fit_transform(dataset)

# Applying the OneHot encoder for blood pressure
OneHot = ce.OneHotEncoder(cols=['BP'])
dataset = OneHot.fit_transform(dataset)

#Applaying binary encoder for Cholesterol
binary_encoder = ce.BinaryEncoder(cols=['Cholesterol'])
dataset = binary_encoder.fit_transform(dataset)

#spliting variables into depended and target value
X = dataset.iloc[:, :-1].values
y = dataset.iloc[:, -1].values

# Applying label encoding to the target variable 'Drug'
label_encoder = LabelEncoder()
y = label_encoder.fit_transform(y)

## Splitting the dataset into the Training set and Test set

In [763]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.4)

In [764]:
print(X_train)

[[47.     1.     0.    ...  0.     1.    13.093]
 [24.     0.     1.    ...  0.     1.    10.605]
 [28.     0.     1.    ...  0.     1.    12.879]
 ...
 [54.     1.     0.    ...  0.     1.    24.658]
 [28.     0.     1.    ...  1.     0.    18.809]
 [35.     0.     1.    ...  0.     1.    12.894]]


In [765]:
print(y_train)

[3 4 4 3 4 3 0 4 3 0 1 4 0 4 0 0 0 0 0 0 0 0 4 0 0 1 2 4 4 3 2 0 0 0 4 4 0
 4 4 3 2 4 2 0 0 0 2 0 0 0 1 0 0 0 0 1 0 2 1 0 2 0 2 4 4 4 0 4 0 4 1 3 4 0
 0 0 4 0 4 0 4 2 0 0 0 0 4 4 0 3 4 0 0 2 0 0 4 0 0 0 4 2 3 1 0 0 4 1 1 1 4
 3 0 4 0 4 0 0 0 1]


In [766]:
print(X_test)

[[66.     1.     0.     1.     0.     0.     0.     1.    16.347]
 [36.     0.     1.     0.     0.     1.     0.     1.    16.753]
 [49.     0.     1.     0.     0.     1.     0.     1.    16.275]
 [57.     0.     1.     0.     0.     1.     1.     0.    25.893]
 [19.     0.     1.     1.     0.     0.     1.     0.    25.969]
 [74.     1.     0.     1.     0.     0.     1.     0.    15.436]
 [21.     0.     1.     1.     0.     0.     1.     0.    28.632]
 [24.     1.     0.     0.     0.     1.     0.     1.    25.786]
 [46.     1.     0.     0.     0.     1.     1.     0.     7.285]
 [63.     1.     0.     0.     0.     1.     0.     1.    25.917]
 [18.     0.     1.     0.     0.     1.     1.     0.     8.75 ]
 [58.     0.     1.     0.     1.     0.     0.     1.    38.247]
 [42.     1.     0.     0.     1.     0.     0.     1.    20.013]
 [28.     0.     1.     0.     0.     1.     0.     1.    19.675]
 [36.     1.     0.     0.     1.     0.     1.     0.    11.424]
 [37.     

In [767]:
print(y_test)

[0 0 0 0 0 0 0 0 4 0 4 0 0 0 4 0 2 4 3 3 3 1 2 1 0 0 1 1 4 0 1 0 4 4 4 1 4
 2 4 1 0 4 0 0 0 1 3 4 4 0 2 0 4 1 0 0 3 4 4 4 2 1 4 0 0 1 4 0 0 0 4 1 0 4
 0 0 0 0 3 4]


## Feature Scaling

In [768]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

In [769]:
print(X_train)

[[ 0.17629875  0.90453403 -0.90453403 ... -0.95118973  0.95118973
  -0.42941879]
 [-1.18668316 -1.1055416   1.1055416  ... -0.95118973  0.95118973
  -0.75789009]
 [-0.94964283 -1.1055416   1.1055416  ... -0.95118973  0.95118973
  -0.45767155]
 ...
 [ 0.59111933  0.90453403 -0.90453403 ... -0.95118973  0.95118973
   1.09741825]
 [-0.94964283 -1.1055416   1.1055416  ...  1.05131497 -1.05131497
   0.32522025]
 [-0.53482225 -1.1055416   1.1055416  ... -0.95118973  0.95118973
  -0.45569122]]


In [770]:
print(X_test)

[[ 1.30224032e+00  9.04534034e-01 -9.04534034e-01  1.20373568e+00
  -7.20407761e-01 -5.77350269e-01 -9.51189731e-01  9.51189731e-01
   1.81530561e-04]
 [-4.75562166e-01 -1.10554160e+00  1.10554160e+00 -8.30747161e-01
  -7.20407761e-01  1.73205081e+00 -9.51189731e-01  9.51189731e-01
   5.37825543e-02]
 [ 2.94818913e-01 -1.10554160e+00  1.10554160e+00 -8.30747161e-01
  -7.20407761e-01  1.73205081e+00 -9.51189731e-01  9.51189731e-01
  -9.32406971e-03]
 [ 7.68899577e-01 -1.10554160e+00  1.10554160e+00 -8.30747161e-01
  -7.20407761e-01  1.73205081e+00  1.05131497e+00 -1.05131497e+00
   1.26046570e+00]
 [-1.48298358e+00 -1.10554160e+00  1.10554160e+00  1.20373568e+00
  -7.20407761e-01 -5.77350269e-01  1.05131497e+00 -1.05131497e+00
   1.27049939e+00]
 [ 1.77632099e+00  9.04534034e-01 -9.04534034e-01  1.20373568e+00
  -7.20407761e-01 -5.77350269e-01  1.05131497e+00 -1.05131497e+00
  -1.20090717e-01]
 [-1.36446341e+00 -1.10554160e+00  1.10554160e+00  1.20373568e+00
  -7.20407761e-01 -5.7735026

##Applying PCA


In [771]:
from sklearn.decomposition import KernelPCA
kpca = KernelPCA(n_components = 2, kernel = 'rbf')
X_train = kpca.fit_transform(X_train)
X_test = kpca.transform(X_test)

## Training the SVM model on the Training set

In [772]:
from sklearn.svm import SVC
classifier = SVC(kernel = 'linear', random_state=0)
classifier.fit(X_train, y_train)

SVC(kernel='linear', random_state=0)

## Predicting the Test set results

In [773]:
y_pred = classifier.predict(X_test)
print(np.concatenate((y_pred.reshape(len(y_pred),1), y_test.reshape(len(y_test),1)),1))

[[0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 4]
 [0 0]
 [0 4]
 [0 0]
 [0 0]
 [0 0]
 [0 4]
 [0 0]
 [0 2]
 [0 4]
 [0 3]
 [0 3]
 [0 3]
 [0 1]
 [0 2]
 [0 1]
 [0 0]
 [0 0]
 [0 1]
 [0 1]
 [0 4]
 [0 0]
 [0 1]
 [0 0]
 [0 4]
 [0 4]
 [0 4]
 [0 1]
 [0 4]
 [0 2]
 [0 4]
 [0 1]
 [0 0]
 [0 4]
 [0 0]
 [0 0]
 [0 0]
 [0 1]
 [0 3]
 [0 4]
 [0 4]
 [0 0]
 [0 2]
 [0 0]
 [0 4]
 [0 1]
 [0 0]
 [0 0]
 [0 3]
 [0 4]
 [0 4]
 [0 4]
 [0 2]
 [0 1]
 [0 4]
 [0 0]
 [0 0]
 [0 1]
 [0 4]
 [0 0]
 [0 0]
 [0 0]
 [0 4]
 [0 1]
 [0 0]
 [0 4]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 3]
 [0 4]]


## Making the Confusion Matrix

In [774]:
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_test, y_pred)
print(cm)
accuracy_score(y_test, y_pred)

[[35  0  0  0  0]
 [12  0  0  0  0]
 [ 5  0  0  0  0]
 [ 6  0  0  0  0]
 [22  0  0  0  0]]


0.4375

# Trying Different number of PCA components

## Importing the dataset and encoding categorical values


In [775]:
dataset = pd.read_csv('drug200.csv')

# Applying the binary encoder for SEX
binary_encoder = ce.BinaryEncoder(cols=['Sex'])
dataset = binary_encoder.fit_transform(dataset)

# Applying the OneHot encoder for blood pressure
OneHot = ce.OneHotEncoder(cols=['BP'])
dataset = OneHot.fit_transform(dataset)

#Applaying binary encoder for Cholesterol
binary_encoder = ce.BinaryEncoder(cols=['Cholesterol'])
dataset = binary_encoder.fit_transform(dataset)

#spliting variables into depended and target value
X = dataset.iloc[:, :-1].values
y = dataset.iloc[:, -1].values

# Applying label encoding to the target variable 'Drug'
label_encoder = LabelEncoder()
y = label_encoder.fit_transform(y)

## Splitting the dataset into the Training set and Test set

In [776]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.4)

In [777]:
print(X_train)

[[31.     1.     0.    ...  1.     0.    11.227]
 [43.     1.     0.    ...  0.     1.    13.972]
 [16.     0.     1.    ...  1.     0.    15.516]
 ...
 [15.     1.     0.    ...  0.     1.     9.084]
 [49.     0.     1.    ...  1.     0.     9.381]
 [38.     1.     0.    ...  0.     1.    18.295]]


In [778]:
print(y_train)

[1 1 0 0 0 0 2 3 4 1 4 0 0 0 0 0 4 0 0 4 2 2 0 4 1 0 0 4 0 0 0 2 0 0 0 0 4
 2 4 4 3 1 2 0 4 0 0 2 4 0 1 0 0 4 4 4 0 1 0 4 2 0 1 1 1 0 2 0 3 3 1 4 1 0
 0 4 0 4 0 4 1 0 0 0 0 4 0 4 0 3 0 3 0 4 0 0 4 1 0 4 0 4 4 3 2 0 0 4 0 0 4
 3 3 0 4 4 4 4 4 0]


In [779]:
print(X_test)

[[48.     0.     1.     0.     1.     0.     0.     1.    15.036]
 [34.     1.     0.     1.     0.     0.     0.     1.    18.703]
 [36.     1.     0.     0.     1.     0.     1.     0.    11.424]
 [47.     1.     0.     0.     1.     0.     0.     1.    13.093]
 [39.     1.     0.     0.     1.     0.     1.     0.    13.938]
 [22.     1.     0.     0.     0.     1.     0.     1.    11.953]
 [49.     0.     1.     0.     0.     1.     0.     1.    16.275]
 [65.     0.     1.     0.     1.     0.     1.     0.    13.769]
 [47.     0.     1.     0.     0.     1.     1.     0.     6.683]
 [49.     1.     0.     1.     0.     0.     1.     0.     8.7  ]
 [49.     1.     0.     1.     0.     0.     1.     0.     6.269]
 [72.     0.     1.     0.     1.     0.     1.     0.    14.642]
 [23.     1.     0.     0.     0.     1.     0.     1.    12.26 ]
 [49.     1.     0.     0.     1.     0.     1.     0.    11.014]
 [36.     0.     1.     1.     0.     0.     0.     1.    11.198]
 [39.     

In [780]:
print(y_test)

[0 0 4 3 4 4 0 4 4 1 1 4 4 4 1 0 4 1 3 0 1 4 0 2 4 3 2 0 3 4 0 4 0 3 0 3 0
 0 0 2 0 1 4 0 0 0 0 0 2 4 1 2 0 0 0 1 0 1 4 4 4 0 0 0 0 0 4 0 0 0 0 0 4 0
 0 3 4 0 0 2]


## Feature Scaling

In [781]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

In [782]:
print(X_train)

[[-0.78670942  0.95118973 -0.95118973 ...  1.0168079  -1.0168079
  -0.67976656]
 [-0.07016194  0.95118973 -0.95118973 ... -0.98346994  0.98346994
  -0.29338227]
 [-1.68239378 -1.05131497  1.05131497 ...  1.0168079  -1.0168079
  -0.07604991]
 ...
 [-1.74210607  0.95118973 -0.95118973 ... -0.98346994  0.98346994
  -0.98141375]
 [ 0.2881118  -1.05131497  1.05131497 ...  1.0168079  -1.0168079
  -0.93960823]
 [-0.36872339  0.95118973 -0.95118973 ... -0.98346994  0.98346994
   0.3151202 ]]


In [783]:
print(X_test)

[[ 0.22839951 -1.05131497  1.05131497 -0.83074716  1.59041245 -0.66766993
  -0.98346994  0.98346994 -0.14361437]
 [-0.60757255  0.95118973 -0.95118973  1.20373568 -0.62876771 -0.66766993
  -0.98346994  0.98346994  0.37254999]
 [-0.48814797  0.95118973 -0.95118973 -0.83074716  1.59041245 -0.66766993
   1.0168079  -1.0168079  -0.65203698]
 [ 0.16868722  0.95118973 -0.95118973 -0.83074716  1.59041245 -0.66766993
  -0.98346994  0.98346994 -0.4171097 ]
 [-0.3090111   0.95118973 -0.95118973 -0.83074716  1.59041245 -0.66766993
   1.0168079  -1.0168079  -0.29816809]
 [-1.32412004  0.95118973 -0.95118973 -0.83074716 -0.62876771  1.49774605
  -0.98346994  0.98346994 -0.5775753 ]
 [ 0.2881118  -1.05131497  1.05131497 -0.83074716 -0.62876771  1.49774605
  -0.98346994  0.98346994  0.03078641]
 [ 1.24350844 -1.05131497  1.05131497 -0.83074716  1.59041245 -0.66766993
   1.0168079  -1.0168079  -0.32195641]
 [ 0.16868722 -1.05131497  1.05131497 -0.83074716 -0.62876771  1.49774605
   1.0168079  -1.01680

##Applying PCA


In [784]:
from sklearn.decomposition import KernelPCA
kpca = KernelPCA(n_components = 6, kernel = 'rbf')
X_train = kpca.fit_transform(X_train)
X_test = kpca.transform(X_test)

## Training the SVM model on the Training set

In [785]:
from sklearn.svm import SVC
classifier = SVC(kernel = 'linear', random_state=0)
classifier.fit(X_train, y_train)

SVC(kernel='linear', random_state=0)

## Predicting the Test set results

In [786]:
y_pred = classifier.predict(X_test)
print(np.concatenate((y_pred.reshape(len(y_pred),1), y_test.reshape(len(y_test),1)),1))

[[3 0]
 [1 0]
 [0 4]
 [0 3]
 [0 4]
 [0 4]
 [4 0]
 [0 4]
 [4 4]
 [1 1]
 [1 1]
 [0 4]
 [0 4]
 [0 4]
 [0 1]
 [4 0]
 [4 4]
 [0 1]
 [0 3]
 [0 0]
 [0 1]
 [0 4]
 [0 0]
 [0 2]
 [4 4]
 [0 3]
 [1 2]
 [0 0]
 [3 3]
 [4 4]
 [0 0]
 [4 4]
 [4 0]
 [3 3]
 [0 0]
 [0 3]
 [1 0]
 [0 0]
 [0 0]
 [1 2]
 [1 0]
 [1 1]
 [4 4]
 [0 0]
 [4 0]
 [0 0]
 [0 0]
 [0 0]
 [1 2]
 [4 4]
 [0 1]
 [1 2]
 [4 0]
 [0 0]
 [0 0]
 [1 1]
 [4 0]
 [1 1]
 [0 4]
 [0 4]
 [0 4]
 [0 0]
 [0 0]
 [0 0]
 [4 0]
 [0 0]
 [4 4]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [4 4]
 [1 0]
 [0 0]
 [0 3]
 [0 4]
 [0 0]
 [4 0]
 [1 2]]


## Making the Confusion Matrix

In [787]:
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_test, y_pred)
print(cm)
accuracy_score(y_test, y_pred)

[[24  4  0  1  8]
 [ 4  5  0  0  0]
 [ 1  5  0  0  0]
 [ 5  0  0  2  0]
 [12  0  0  0  9]]


0.5

# Applying the other Models (KNN, NCC)

## Importing the dataset and encoding categorical values


In [788]:
dataset = pd.read_csv('drug200.csv')

# Applying the binary encoder for SEX
binary_encoder = ce.BinaryEncoder(cols=['Sex'])
dataset = binary_encoder.fit_transform(dataset)

# Applying the OneHot encoder for blood pressure
OneHot = ce.OneHotEncoder(cols=['BP'])
dataset = OneHot.fit_transform(dataset)

#Applaying binary encoder for Cholesterol
binary_encoder = ce.BinaryEncoder(cols=['Cholesterol'])
dataset = binary_encoder.fit_transform(dataset)

#spliting variables into depended and target value
X = dataset.iloc[:, :-1].values
y = dataset.iloc[:, -1].values

# Applying label encoding to the target variable 'Drug'
label_encoder = LabelEncoder()
y = label_encoder.fit_transform(y)

## Splitting the dataset into the Training set and Test set

In [789]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.4)

In [790]:
print(X_train)

[[23.     0.     1.    ...  0.     1.    25.355]
 [23.     1.     0.    ...  0.     1.     8.011]
 [67.     1.     0.    ...  1.     0.    20.693]
 ...
 [37.     0.     1.    ...  1.     0.    12.006]
 [43.     1.     0.    ...  1.     0.    19.368]
 [31.     1.     0.    ...  0.     1.    30.366]]


In [791]:
print(y_train)

[0 1 0 4 0 0 0 0 3 0 0 2 0 4 4 0 1 2 4 1 0 4 0 0 4 2 4 1 4 3 3 3 1 3 4 0 4
 0 4 0 2 4 3 4 4 0 1 0 1 4 4 0 4 0 0 0 0 0 2 0 4 0 2 0 0 0 4 0 0 0 0 1 4 3
 4 0 2 1 2 0 0 0 4 1 0 0 3 4 0 4 0 0 4 4 0 4 1 0 0 0 4 0 1 0 4 0 0 0 4 4 4
 4 2 4 0 4 0 4 0 0]


In [792]:
print(X_test)

[[58.     1.     0.     1.     0.     0.     0.     1.    18.991]
 [50.     0.     1.     0.     0.     1.     0.     1.    12.703]
 [72.     0.     1.     0.     1.     0.     1.     0.    14.642]
 [24.     0.     1.     0.     0.     1.     0.     1.    10.605]
 [47.     1.     0.     1.     0.     0.     0.     1.    10.403]
 [33.     0.     1.     0.     1.     0.     0.     1.    33.486]
 [26.     0.     1.     1.     0.     0.     1.     0.    12.307]
 [53.     1.     0.     0.     0.     1.     0.     1.    14.133]
 [39.     1.     0.     0.     1.     0.     1.     0.    13.938]
 [56.     0.     1.     1.     0.     0.     0.     1.    25.395]
 [28.     0.     1.     1.     0.     0.     1.     0.    18.809]
 [32.     0.     1.     1.     0.     0.     1.     0.    10.292]
 [74.     0.     1.     0.     1.     0.     0.     1.    20.942]
 [64.     0.     1.     0.     1.     0.     1.     0.    25.741]
 [59.     0.     1.     0.     1.     0.     0.     1.    10.444]
 [61.     

In [793]:
print(y_test)

[0 4 4 4 1 0 1 4 4 0 0 1 0 0 3 0 4 3 4 0 0 0 4 0 4 3 0 3 0 1 0 1 0 2 0 0 0
 3 4 0 0 0 2 0 4 2 0 4 0 1 0 4 3 4 3 2 0 1 2 0 0 0 2 4 2 0 3 0 4 0 1 4 0 0
 1 4 1 1 0 0]


## Feature Scaling

In [794]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

In [795]:
print(X_train)

[[-1.27185288 -1.1055416   1.1055416  ... -1.06904497  1.06904497
   1.29833232]
 [-1.27185288  0.90453403 -0.90453403 ... -1.06904497  1.06904497
  -1.131184  ]
 [ 1.36163073  0.90453403 -0.90453403 ...  0.93541435 -0.93541435
   0.64528774]
 ...
 [-0.43392628 -1.1055416   1.1055416  ...  0.93541435 -0.93541435
  -0.57157159]
 [-0.07481488  0.90453403 -0.90453403 ...  0.93541435 -0.93541435
   0.45968413]
 [-0.79303768  0.90453403 -0.90453403 ... -1.06904497  1.06904497
   2.00026418]]


In [796]:
print(X_test)

[[ 0.82296363  0.90453403 -0.90453403  1.24627103 -0.65465367 -0.66766993
  -1.06904497  1.06904497  0.40687464]
 [ 0.34414843 -1.1055416   1.1055416  -0.80239368 -0.65465367  1.49774605
  -1.06904497  1.06904497 -0.47393708]
 [ 1.66089023 -1.1055416   1.1055416  -0.80239368  1.52752523 -0.66766993
   0.93541435 -0.93541435 -0.20232545]
 [-1.21200098 -1.1055416   1.1055416  -0.80239368 -0.65465367  1.49774605
  -1.06904497  1.06904497 -0.76782115]
 [ 0.16459273  0.90453403 -0.90453403  1.24627103 -0.65465367 -0.66766993
  -1.06904497  1.06904497 -0.79611695]
 [-0.67333388 -1.1055416   1.1055416  -0.80239368  1.52752523 -0.66766993
  -1.06904497  1.06904497  2.43730817]
 [-1.09229718 -1.1055416   1.1055416   1.24627103 -0.65465367 -0.66766993
   0.93541435 -0.93541435 -0.52940805]
 [ 0.52370413  0.90453403 -0.90453403 -0.80239368 -0.65465367  1.49774605
  -1.06904497  1.06904497 -0.27362526]
 [-0.31422248  0.90453403 -0.90453403 -0.80239368  1.52752523 -0.66766993
   0.93541435 -0.93541

## Training the KNN model on the Training set

In [797]:
from sklearn.neighbors import KNeighborsClassifier
classifierKNN = KNeighborsClassifier(n_neighbors = 5, metric = 'minkowski', p = 2)
classifierKNN.fit(X_train, y_train)

KNeighborsClassifier()

##Training the Nearest Class Centroid model

In [798]:
from sklearn.neighbors import NearestCentroid
classifierNC = NearestCentroid(metric='euclidean')
classifierNC.fit(X_train, y_train)

NearestCentroid()

## Predicting the Test set results

In [799]:
y_pred_KNN = classifierKNN.predict(X_test)
y_pred_NC = classifierNC.predict(X_test)

## Making the Confusion Matrix

In [800]:
from sklearn.metrics import accuracy_score
print("The accurasy score of KNN Model is: ", accuracy_score(y_test, y_pred_KNN) )
print("The accurasy score of NCC Model is: ", accuracy_score(y_test, y_pred_NC) )

The accurasy score of KNN Model is:  0.7
The accurasy score of KNN Model is:  0.7625


#Documendation


## Introduction

- **The Dataset:** https://www.kaggle.com/datasets/prathamtripathi/drug-classification/data
- **Objectives of the analysis:** I tryied several methodes to find the best "doctor" model, so it can predict based on the features what type of drug a patient needs.

# Methodology

### Preprocessing Steps

1. **Encoding Categorical Values**:
   - **Binary Encoding**: Applied to the 'Sex' and 'Cholesterol' columns.
   - **One-Hot Encoding**: Used for the 'BP' (Blood Pressure) column.

2. **Feature Scaling**:
   - **StandardScaler**: Standardizes features by removing the mean and scaling to unit variance.

### Parameter Selection and Model Tuning Methods

1. **Support Vector Machine (SVM)**:
   - **Kernel**: Linear
   - **Random State**: 0

2. **K-Nearest Neighbors (KNN)**:
   - **Number of Neighbors (n_neighbors)**: 5
   - **Distance Metric**: Minkowski with \( p = 2 \) (Euclidean distance)

3. **Nearest Class Centroid (NCC)**:
   - **Metric**: Euclidean distance

### Cross-Validation Strategy

- **K-Fold Cross-Validation**: Different values of `n_splits` (2, 10, 15, 25) are experimented with.
- **Shuffle**: Ensuring data is shuffled before splitting into batches.

### Additional Points

- **Principal Component Analysis (PCA)**: Used to reduce dimensionality.
- **Model Comparison**: Comparing SVM with KNN and NCC models.


# Algorithm Explanation

### Support Vector Machine (SVM)

- **Fundamentals**: SVM is a supervised learning algorithm used for classification and regression. It works by finding the hyperplane that best separates the classes in the feature space.
- **Maximizing Margin**: The algorithm focuses on maximizing the margin between the data points of different classes, aiming to increase the model's generalization ability.
- **Support Vectors**: Data points closest to the hyperplane, called support vectors, are critical in defining the hyperplane and the margin.

### Nearest Neighbor (KNN)

- **Basic Concept**: Classifies a data point based on how its neighbors are classified.
- **Distance Metrics**: Often uses Euclidean distance but can use other metrics.
- **Choosing 'K'**: The number of neighbors (K) is a crucial parameter, with a larger K smoothing the decision boundaries.

### Nearest Class Centroid (NCC)

- **Principle**: Assigns a class based on the closest class centroid.
- **Centroid Calculation**: The centroid of a class is the average of all points in that class.
- **Suitability**: Effective when classes are compact and well-separated.


# Results

- **Simple Support Vector Machine (SVM)**: accuracy is `0.9375`
- **Simple Support Vector Machine (SVM) with cross validation:** accuracy varies from `0.9901098` to `0.985` based on the number of splits
- **Simple Support Vector Machine (SVM) with PCA Applayied:** accuracy varies based on the number of PCA  components with the highest `0.5`
- **Comparative analysis with Nearest Neighbor and Nearest Class Centroid classifiers:** For NC accuracy is `0.7625` and for KNN `0.7`

Since I didn't initialize a random state, results might slightly vary from time to time.

# Discussion

The results indicate a strong performance of the Simple Support Vector Machine (SVM) model, achieving an accuracy of `0.9375`. This high accuracy underscores SVM's effectiveness in handling the dataset's linear separability. Notably, the application of cross-validation to SVM further enhanced its robustness, with accuracy reaching as high as `0.9901098`, depending on the number of splits. This variability in accuracy with different splits highlights the importance of cross-validation in evaluating model performance more comprehensively. However, the introduction of PCA (Principal Component Analysis) to SVM led to a significant decrease in accuracy, with the highest being only `0.5`. This suggests that dimensionality reduction via PCA might be removing critical information necessary for accurate classification in this specific dataset. In comparison, Nearest Neighbor (KNN) and Nearest Class Centroid (NCC) classifiers demonstrated lower accuracy, with NCC achieving 0.7625 and KNN 0.7. This disparity in performance could be attributed to SVM's superior ability to manage the feature space and its boundaries, as opposed to the distance-based classification approach used by KNN and the centroid-based approach of NCC.

